### DISCUSSION TITLE
Generating molecules with chiral tags and arbitrary atom/bond orders

### DISCUSSION THREAD

**Discussion by najwalb** [rdkit/rdkit#7421] [Q&A]
My question is similar to [this issue](https://github.com/rdkit/rdkit/issues/2736) but perhaps broader in scope. I am generating molecules as 2D graphs, meaning with a permutation equivariant model producing atoms/bonds in an arbitrary order. I would like to add chiral tags to this graph structure in a consistent way since I need to compare the molecule to a ground truth example and I need the two to match.

From the issue linked above I understand that the chiral tags in rdkit are dependent on bond orders and that there is no straightforward way around this. Based on this I thought of the following two solutions: 1) enforce some canonicalization before adding the chiral tags, 2) define a canonical form using the CIP algorithm. I tried the first approach but it turns out it's not possible because chiral tags change the order of bonds/atoms. Here is an example to illustrate:

```
smi = '[CH3:2][c:3]1[n:4][cH:5][cH:6][n:7]1-[c:8]1[cH:9][cH:10][c:11]([NH:12][c:13]2[n:14][c:15]3[c:16]([c:17]([NH:18][CH2:19][C@@H:20]4[CH2:21][CH2:22][CH2:23][O:24]4)[n:25]2)[CH2:26][NH:27][CH2:28][CH2:29]3)[cH:30][cH:31]1'
orig_mol = Chem.MolFromSmiles(smi)
[a.SetAtomMapNum(0) for a in orig_mol.GetAtoms()]

# canonicalize the molecule to force a bond/atom order
orig_mol = Chem.MolFromSmiles(Chem.MolToSmiles(orig_mol, canonical=True))

# my application requires kekulization for all molecules
Chem.Kekulize(orig_mol, clearAromaticFlags=True)

# get atom symbols
atom_symbols = [atom.GetSymbol() for atom in orig_mol.GetAtoms()]
# get atom charges
atom_charges = [atom.GetFormalCharge() for atom in orig_mol.GetAtoms()]
# get atom chiral tags
chiral_tags = [atom.GetChiralTag() for atom in orig_mol.GetAtoms()]
# get bond types
bond_types = [(bond.GetBeginAtomIdx(), bond.GetEndAtomIdx(), bond.GetBondType()) for bond in orig_mol.GetBonds()]
# get bond dirs
bond_dirs = [(bond.GetBeginAtomIdx(), bond.GetEndAtomIdx(), bond.GetBondDir()) for bond in orig_mol.GetBonds()]

# show the atom index and chiral tag of any atom with chiral info
chiral_not_none = [(a.GetIdx(), a.GetChiralTag()) for a in orig_mol.GetAtoms() if a.GetChiralTag()!=Chem.rdchem.ChiralType.CHI_UNSPECIFIED]
print(f'orig_mol: chiral_not_none {chiral_not_none}\n')

# reconstruct the molecule
rw_mol = Chem.RWMol()

# add atoms
for i in range(len(atom_symbols)):
    atom = Chem.Atom(atom_symbols[i])
    atom.SetFormalCharge(atom_charges[i])
    molIdx = rw_mol.AddAtom(atom)

# add bonds
for i, (begatom, endatom, btype) in enumerate(bond_types):
    rw_mol.AddBond(begatom, endatom, btype)
    rw_mol.GetBondBetweenAtoms(begatom, endatom).SetBondDir(bond_dirs[i][-1])

# extract molecule object
new_mol = rw_mol.GetMol()
# canonicalize molecule
new_mol = Chem.MolFromSmiles(Chem.MolToSmiles(new_mol, canonical=True))
# kekulize
Chem.Kekulize(new_mol, clearAromaticFlags=True)

# add chiral tags
for atom in new_mol.GetAtoms():
    if chiral_tags[atom.GetIdx()]!=Chem.rdchem.ChiralType.CHI_UNSPECIFIED:
        print(f'rxn {rxn_nb} chiral_tags[{atom.GetIdx()}] {chiral_tags[atom.GetIdx()]}\n')
        atom.SetChiralTag(chiral_tags[atom.GetIdx()])
        print(f'chiral_tags[{atom.GetIdx()}] {atom.GetChiralTag()}\n') # atom idx should be the same as orig_mol: chiral_not_none

# print the new molecule without canonicalization
new_smi_no_cano = Chem.MolToSmiles(new_mol, canonical=True)
print(f'new_smi_no_cano {new_smi_no_cano}\n') # notice that the chiral tag is added to a different atom even if it had the same index
    
# canonicalize both molecules
orig_mol = Chem.MolFromSmiles(Chem.MolToSmiles(orig_mol, canonical=True))
new_mol = Chem.MolFromSmiles(Chem.MolToSmiles(new_mol, canonical=True))

new_smi = Chem.MolToSmiles(new_mol, canonical=True)
orig_smi = Chem.MolToSmiles(orig_mol, canonical=True)

if orig_smi!=new_smi:
    print(f'rxn {rxn_nb}\n')
    print(f'Original smiles: {orig_smi}\n')
    print(f'Rebuilt smiles: {new_smi}\n') # the chiral tag disappears here because it was not in the right place
    print(f'Original smiles==Rebuilt smiles: {orig_smi==new_smi}\n')
```

Basically the issue with canonicalization is that there is no way to use the canonical smiles order while keeping track of where the chiral tag should be put (and the canonicalization is different with and without chiral tags). The second option is to use an approach similar to what @greglandrum proposed in [this notebook](https://gist.github.com/greglandrum/9f0e068e53171174b6797348eca64b3e) but with CIP weights replacing neighboring atom index. Is it possible to get CIP weights using some rdkit functions?

Any other feedback/approaches are more than welcome.

**Comment by bp-kelley** [comment]
These lines are changing your atom orders (canonicalization can changes indices)

```
# extract molecule object
new_mol = rw_mol.GetMol()
# canonicalize moleculenew_mol = Chem.MolFromSmiles(Chem.MolToSmiles(new_mol, canonical=True))
# kekulize
Chem.Kekulize(new_mol, clearAromaticFlags=True)
```

Note that a "Chem.RWMol" is a valid molecule, so making the following change fixes your issue:

```
new_mol = rw_mol # an rw_mol IS a molecule
# kekulize
Chem.Kekulize(new_mol, clearAromaticFlags=True)
```

Hope this helps.

**Reply by najwalb** [reply]
Hi thanks for the suggestion, but actually the canonicalization was used on purpose to try to fix the atom/bond order when recreating the molecule. I agree that if I input the atoms/bonds to the rwMol in the same order I read them out from orig_mol, and don't use canonicalization, I am able to recover the SMILES string I started with. But in my application, the atom/bond orders will be shuffled in all cases, so I tried using canonicalization to 'reorder' orig_mol and rwMol in the same way before adding chiral tags.

Here is an example with shuffled atoms/bonds where I fail to recover the smiles with/without canonicalization:

```
smi = 'C[O:1][C:2]([c:3]1[cH:4][cH:5][c:6](/[CH:7]=[CH:8]/[C:9]2=[CH:10][CH2:11][C@:12]3([CH3:13])[C@H:14]4[CH2:15][CH2:16][C@H:17]5[C@@:18]([CH3:19])([CH2:20][CH2:21][C@@:22]6([NH:23][CH2:24][CH2:25][N:26]7[CH2:27][CH2:28][S:29](=[O:30])(=[O:31])[CH2:32][CH2:33]7)[CH2:34][CH2:35][C@@H:36]([C:37](=[CH2:38])[CH3:39])[C@H:40]56)[C@:41]4([CH3:42])[CH2:43][CH2:44][C@H:45]3[C:46]2([CH3:47])[CH3:48])[cH:49][cH:50]1)=[O:51]'
orig_mol = Chem.MolFromSmiles(smi)
[a.SetAtomMapNum(0) for a in orig_mol.GetAtoms()]
orig_mol = Chem.MolFromSmiles(Chem.MolToSmiles(orig_mol, canonical=True))
Chem.Kekulize(orig_mol, clearAromaticFlags=True)

# get atom symbols
atom_symbols = [atom.GetSymbol() for atom in orig_mol.GetAtoms()]
# get atom charges
atom_charges = [atom.GetFormalCharge() for atom in orig_mol.GetAtoms()]
# get atom chiral tags
chiral_tags = [atom.GetChiralTag() for atom in orig_mol.GetAtoms()]
# get bond types
bond_types = [(bond.GetBeginAtomIdx(), bond.GetEndAtomIdx(), bond.GetBondType()) for bond in orig_mol.GetBonds()]
# get bond dirs
bond_dirs = [(bond.GetBeginAtomIdx(), bond.GetEndAtomIdx(), bond.GetBondDir()) for bond in orig_mol.GetBonds()]

chiral_not_none = [(a.GetIdx(), a.GetChiralTag()) for a in orig_mol.GetAtoms() if a.GetChiralTag()!=Chem.rdchem.ChiralType.CHI_UNSPECIFIED]
#print(f'rxn {rxn_nb} orig_mol: chiral_not_none {chiral_not_none}\n')

# shuffle bond types
torch.manual_seed(0)
perm = torch.randperm(len(bond_types)).tolist()
shuffled_bond_types = {(bond_types[i][0], bond_types[i][1]):bond_types[i][2] for i in perm}
shuffle_bond_dirs = {(bond_types[i][0], bond_types[i][1]):bond_dirs[i][2] for i in perm}

perm = torch.randperm(len(atom_symbols))
atom_symbols = [atom_symbols[i] for i in perm]
atom_charges = [atom_charges[i] for i in perm]
chiral_tags = [chiral_tags[i] for i in perm]
perm_inv = torch.argsort(perm)
perm_edges = torch.randperm(len(bond_types))
bond_types = [(perm_inv[bond_types[e][0]].item(), perm_inv[bond_types[e][1]].item(), bond_types[e][2]) for e in perm_edges]
bond_dirs = [(perm_inv[bond_dirs[e][0]].item(), perm_inv[bond_dirs[e][1]].item(), bond_dirs[e][2]) for e in perm_edges]
        
rw_mol = Chem.RWMol()

# add atoms
for i in range(len(atom_symbols)):
    atom = Chem.Atom(atom_symbols[i])
    atom.SetFormalCharge(atom_charges[i])
    molIdx = rw_mol.AddAtom(atom)

for i, (begatom, endatom, btype) in enumerate(bond_types):
    rw_mol.AddBond(begatom, endatom, btype)
    rw_mol.GetBondBetweenAtoms(begatom, endatom).SetBondDir(bond_dirs[i][-1])
    
new_mol = rw_mol
# COMMENT THE LINE BELOW OUT TO TRY WITHOUT CANONICALIZATION
new_mol = Chem.MolFromSmiles(Chem.MolToSmiles(new_mol, canonical=True))
Chem.Kekulize(new_mol, clearAromaticFlags=True)

for atom in new_mol.GetAtoms():
    if chiral_tags[atom.GetIdx()]!=Chem.rdchem.ChiralType.CHI_UNSPECIFIED:
        #print(f'rxn {rxn_nb} chiral_tags[{atom.GetIdx()}] {chiral_tags[atom.GetIdx()]}\n')
        atom.SetChiralTag(chiral_tags[atom.GetIdx()])
        #print(f'chiral_tags[{atom.GetIdx()}] {atom.GetChiralTag()}\n')

chiral_not_none_new = [(a.GetIdx(), a.GetChiralTag()) for a in new_mol.GetAtoms() if a.GetChiralTag()!=Chem.rdchem.ChiralType.CHI_UNSPECIFIED]
#print(f'rxn {rxn_nb} orig_mol: chiral_not_none_new {chiral_not_none_new}\n')

new_smi_no_cano = Chem.MolToSmiles(new_mol, canonical=True)
#print(f'new_smi_no_cano {new_smi_no_cano}\n')
    
orig_mol = Chem.MolFromSmiles(Chem.MolToSmiles(orig_mol, canonical=True))
new_mol = Chem.MolFromSmiles(Chem.MolToSmiles(new_mol, canonical=True))

new_smi = Chem.MolToSmiles(new_mol, canonical=True)
orig_smi = Chem.MolToSmiles(orig_mol, canonical=True)

if orig_smi!=new_smi:
    print(f'rxn {rxn_nb}\n')
    print(f'Original smiles: {orig_smi}\n')
    print(f'Rebuilt smiles: {new_smi}\n')
    print(f'Original smiles==Rebuilt smiles: {orig_smi==new_smi}\n')
```
